core/lib/zksync_core/src/native_erc20_fetcher/mod.rs

```diff
@@ -0,0 +1,107 @@
+use std::{sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use metrics::atomics::AtomicU64;
+use tokio::{
+    sync::{watch, OnceCell},
+    task::JoinHandle,
+};
+use zksync_config::configs::native_erc20_fetcher::NativeErc20FetcherConfig;
+
+// TODO: this error type is also defined by the gasAdjuster module,
+//       we should consider moving it to a common place
+#[derive(thiserror::Error, Debug, Clone)]
+#[error(transparent)]
+pub struct Error(Arc<anyhow::Error>);
+
+impl From<anyhow::Error> for Error {
+    fn from(err: anyhow::Error) -> Self {
+        Self(Arc::new(err))
+    }
+}
+
+pub trait Erc20Fetcher: 'static + std::fmt::Debug + Send + Sync {
+    fn conversion_rate(&self) -> anyhow::Result<u64>;
+}
+
+pub(crate) struct NativeErc20FetcherSingleton {
+    native_erc20_fetcher_config: NativeErc20FetcherConfig,
+    singleton: OnceCell<Result<Arc<NativeErc20Fetcher>, Error>>,
+}
+
+impl NativeErc20FetcherSingleton {
+    pub fn new(native_erc20_fetcher_config: NativeErc20FetcherConfig) -> Self {
+        Self {
+            native_erc20_fetcher_config,
+            singleton: OnceCell::new(),
+        }
+    }
+
+    pub async fn get_or_init(&mut self) -> Result<Arc<NativeErc20Fetcher>, Error> {
+        let adjuster = self
+            .singleton
+            .get_or_init(|| async {
+                let fetcher = NativeErc20Fetcher::new(self.native_erc20_fetcher_config.clone());
+                Ok(Arc::new(fetcher))
+            })
+            .await;
+        adjuster.clone()
+    }
+
+    pub fn run_if_initialized(
+        self,
+        stop_signal: watch::Receiver<bool>,
+    ) -> Option<JoinHandle<anyhow::Result<()>>> {
+        let fetcher = self.singleton.get()?.clone();
+        Some(tokio::spawn(async move { fetcher?.run(stop_signal).await }))
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct NativeErc20Fetcher {
+    // TODO: we probably need to add a http client here
+    // to avoid creating a new one for each request
+    pub config: NativeErc20FetcherConfig,
+    pub latest_to_eth_conversion_rate: AtomicU64,
+}
+
+impl NativeErc20Fetcher {
+    pub(crate) fn new(config: NativeErc20FetcherConfig) -> Self {
+        Self {
+            config,
+            latest_to_eth_conversion_rate: AtomicU64::new(0),
```

Comment:
You're right, changed here [773ff7a](https://github.com/matter-labs/zksync-era/pull/996/commits/773ff7a2b1977b30cfd192e13073b615ca00522e). Now the conversion rate is fetched when creating a `NativeErc20Fetcher`. Initialization panics if the network request fails, which makes sense since it doesn't make sense to run the network if a reliable conversion rate cannot be first established.

Reply:
I think we can start with that, yes. Although probably later we would like to implement something more robust (on a first glance, running with unsynchronized ratios while the API is being fixed is better than having a couple of hours of downtime for any production environment), we can tackle this separately.